File: python/browserstate/storage/redis_storage.py
import os
import io
import zipfile
import tempfile
import shutil
import logging
import base64
import json
import time
import pathlib
from typing import List

import redis  # Requires: pip install redis

from .storage_provider import StorageProvider

def is_zipfile_safe(zip_file_path: str, target_path: str) -> bool:
    """
    Check if a ZIP file is safe to extract (no directory traversal attacks).
    
    Args:
        zip_file_path: Path to the ZIP file
        target_path: Target extraction directory
        
    Returns:
        True if the ZIP file is safe, False otherwise
    """
    target_path = os.path.normpath(os.path.abspath(target_path))
    
    with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
        for zip_info in zip_ref.infolist():
            # Skip directories
            if zip_info.filename.endswith('/'):
                continue
                
            # Resolve the complete path
            extracted_path = os.path.normpath(os.path.join(target_path, zip_info.filename))
            
            # Check if path would escape the target directory
            if not extracted_path.startswith(target_path):
                return False
    
    return True

def safe_extract_zip(zip_file_path: str, target_path: str) -> None:
    """
    Safely extract a ZIP file, preventing directory traversal attacks.
    
    Args:
        zip_file_path: Path to the ZIP file
        target_path: Target extraction directory
        
    Raises:
        ValueError: If the ZIP file contains unsafe entries
    """
    # Check if ZIP is safe
    if not is_zipfile_safe(zip_file_path, target_path):
        raise ValueError("Security risk: ZIP file contains entries that would extract outside target directory")
    
    # Extract the ZIP file
    with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
        zip_ref.extractall(target_path)

class RedisStorage(StorageProvider):
    """
    Storage provider implementation that uses Redis to store browser sessions
    as compressed ZIP archives to match the TypeScript implementation.
    """
    
    def __init__(self, 
                 redis_url: str = "redis://localhost:6379/0", 
                 key_prefix: str = "browserstate"):
        """
        Initialize RedisStorage provider.
        
        Args:
            redis_url: Redis connection URL.
            key_prefix: Prefix to use for keys in Redis. Must not contain colons.
        """
        # Validate key_prefix format
        if ":" in key_prefix:
            raise ValueError("key_prefix must not contain colons (:). The implementation automatically builds Redis keys in the format: {prefix}:{userId}:{sessionId}")
            
        self.redis_client = redis.Redis.from_url(redis_url)
        self.key_prefix = key_prefix
        logging.info(f"Redis storage initialized with prefix: {self.key_prefix}")
    
    def _get_key(self, user_id: str, session_id: str) -> str:
        """Generate a Redis key for a given user and session."""
        # Validate that user_id and session_id don't contain colons
        if ":" in user_id:
            raise ValueError("user_id must not contain colons (:)")
        if ":" in session_id:
            raise ValueError("session_id must not contain colons (:)")
        return f"{self.key_prefix}:{user_id}:{session_id}"
    
    def _get_metadata_key(self, user_id: str, session_id: str) -> str:
        """Generate a Redis key for session metadata."""
        # Reuse validation from _get_key
        self._get_key(user_id, session_id)
        return f"{self.key_prefix}:{user_id}:{session_id}:metadata"
    
    def _get_temp_path(self, user_id: str, session_id: str) -> str:
        """Get a temporary path for a session."""
        temp_dir = os.path.join(tempfile.gettempdir(), "browserstate", user_id)
        os.makedirs(temp_dir, exist_ok=True)
        return os.path.join(temp_dir, session_id)
    
    def download(self, user_id: str, session_id: str) -> str:
        """
        Downloads a browser session from Redis, extracts the ZIP archive, and writes it
        to a local temporary directory.
        
        Args:
            user_id: User identifier.
            session_id: Session identifier.
            
        Returns:
            Path to the local directory containing the session data.
        """
        key = self._get_key(user_id, session_id)
        metadata_key = self._get_metadata_key(user_id, session_id)
        
        logging.info(f"Looking up session data at Redis key: {key}")
        
        # Get base64-encoded zip data from Redis
        zip_data_base64 = self.redis_client.get(key)
        
        target_path = self._get_temp_path(user_id, session_id)
        
        if os.path.exists(target_path):
            shutil.rmtree(target_path)
        os.makedirs(target_path, exist_ok=True)
        
        if zip_data_base64 is None:
            # No session found; return an empty directory.
            logging.info(f"No session found at key: {key}")
            return target_path
        
        try:
            # Decode base64 data
            logging.info(f"Found session data of size: {len(zip_data_base64)} bytes")
            zip_data = base64.b64decode(zip_data_base64)
            logging.info(f"Decoded base64 data of size: {len(zip_data)} bytes")
            
            # Create temporary zip file
            zip_file_path = os.path.join(
                tempfile.gettempdir(),
                f"{user_id}-{session_id}-{os.getpid()}.zip"
            )
            
            # Write zip data to temporary file
            with open(zip_file_path, 'wb') as f:
                f.write(zip_data)
            
            logging.info(f"Extracting ZIP file to: {target_path}")
            
            # Safely extract zip file to target directory
            safe_extract_zip(zip_file_path, target_path)
                
            # Clean up temporary zip file
            os.remove(zip_file_path)
            
            logging.info(f"Extracted session data to {target_path}")
            
        except Exception as e:
            logging.error(f"Error extracting session from Redis: {e}")
            raise
        
        return target_path
    
    def upload(self, user_id: str, session_id: str, file_path: str) -> None:
        """
        Compresses the session directory into a ZIP archive and uploads it to Redis.
        Uses base64 encoding to match TypeScript implementation.
        
        Args:
            user_id: User identifier.
            session_id: Session identifier.
            file_path: Path to the local directory containing session data.
        """
        key = self._get_key(user_id, session_id)
        metadata_key = self._get_metadata_key(user_id, session_id)
        
        logging.info(f"Uploading session to Redis key: {key}")
        
        # Create temporary zip file
        zip_file_path = os.path.join(
            tempfile.gettempdir(),
            f"{user_id}-{session_id}-{os.getpid()}.zip"
        )
        
        try:
            # Create ZIP archive with maximum compression
            with zipfile.ZipFile(zip_file_path, 'w', zipfile.ZIP_DEFLATED, compresslevel=9) as zipf:
                for root, dirs, files in os.walk(file_path):
                    for file in files:
                        file_path_full = os.path.join(root, file)
                        try:
                            arcname = os.path.relpath(file_path_full, file_path)
                            zipf.write(file_path_full, arcname)
                        except Exception as e:
                            logging.warning(f"Error adding file to ZIP: {file_path_full} - {e}")
            
            # Read zip file as binary
            with open(zip_file_path, 'rb') as f:
                zip_bytes = f.read()
            
            # Get file size for logging
            zip_size = os.path.getsize(zip_file_path)
            logging.info(f"Created ZIP archive of size: {zip_size} bytes")
            
            # Convert to base64 for Redis storage (matching TypeScript implementation)
            zip_base64 = base64.b64encode(zip_bytes)
            logging.info(f"Base64 encoded data size: {len(zip_base64)} bytes")
            
            # Store in Redis
            self.redis_client.set(key, zip_base64)
            
            # Create metadata (matching TypeScript metadata format)
            metadata = {
                "timestamp": time.time() * 1000,  # Current time in milliseconds
                "version": "2.0",
                "encrypted": False,  # Prepare for future encryption support
            }
            
            # Store metadata in Redis
            self.redis_client.set(metadata_key, json.dumps(metadata))
            
            # Clean up temporary zip file
            os.remove(zip_file_path)
            
            logging.info(f"Successfully uploaded session {session_id} to Redis at key: {key}")
            
        except Exception as e:
            logging.error(f"Error uploading session to Redis: {e}")
            
            # Clean up temporary zip file if it exists
            if os.path.exists(zip_file_path):
                os.remove(zip_file_path)
                
            raise
    
    def list_sessions(self, user_id: str) -> List[str]:
        """
        Lists all available sessions for a user from Redis.
        
        Args:
            user_id: User identifier.
            
        Returns:
            List of session identifiers.
        """
        # Validate user_id
        if ":" in user_id:
            raise ValueError("user_id must not contain colons (:)")
            
        pattern = f"{self.key_prefix}:{user_id}:*"
        logging.info(f"Listing sessions with pattern: {pattern}")
        
        try:
            keys = self.redis_client.keys(pattern)
            session_ids = []
            prefix_len = len(f"{self.key_prefix}:{user_id}:")
            
            for key in keys:
                key_str = key.decode('utf-8') if isinstance(key, bytes) else key
                # Extract sessionId from key
                remaining = key_str[prefix_len:]
                
                # Only process keys without additional colons (to exclude metadata)
                if ":" not in remaining and remaining not in session_ids:
                    session_ids.append(remaining)
            
            logging.info(f"Found {len(session_ids)} sessions for user {user_id}")
            return session_ids
        except Exception as e:
            logging.error(f"Error listing sessions in Redis: {e}")
            return []
    
    def delete_session(self, user_id: str, session_id: str) -> None:
        """
        Deletes a browser session from Redis.
        
        Args:
            user_id: User identifier.
            session_id: Session identifier.
        """
        key = self._get_key(user_id, session_id)
        metadata_key = self._get_metadata_key(user_id, session_id)
        logging.info(f"Deleting session at keys: {key}, {metadata_key}")
        
        try:
            # Delete both session data and metadata
            self.redis_client.delete(key, metadata_key)
            logging.info(f"Successfully deleted session {session_id}")
        except Exception as e:
            logging.error(f"Error deleting session from Redis: {e}")
            raise

File: typescript/src/storage/RedisStorage.ts
import { StorageProvider } from "./StorageProvider";
import fs from "fs-extra";
import path from "path";
import os from "os";
import { modules } from "../utils/DynamicImport";

// Define types without importing ioredis directly
interface RedisOptions {
  host: string;
  port: number;
  password?: string;
  db?: number;
  tls?: {
    rejectUnauthorized?: boolean;
    ca?: string[];
    cert?: string;
    key?: string;
  };
  retryStrategy?: (times: number) => number;
  maxRetriesPerRequest?: number;
  enableReadyCheck?: boolean;
}

// Define Redis interface for client operations
interface RedisClient {
  get(key: string): Promise<string | null>;
  set(key: string, value: string): Promise<"OK">;
  setex(key: string, seconds: number, value: string): Promise<"OK">;
  del(key: string): Promise<number>;
  keys(pattern: string): Promise<string[]>;
}

interface SessionMetadata {
  timestamp?: number;
  fileCount?: number;
  version?: string;
  encrypted?: boolean;
}

// Type for metadata created during upload
interface SessionUploadMetadata {
  timestamp: number;
  fileCount: number;
  version: string;
  encrypted: boolean;
}

/**
 * Extract-zip options interface
 */
interface ExtractOptions {
  dir: string;
  onEntry?: (entry: { fileName: string }, zipfile: unknown) => void;
}

/**
 * Redis Storage Architecture
 * =========================
 *
 * ┌─────────────────┐       ┌─────────────────┐        ┌─────────────────────┐
 * │                 │       │                 │        │                     │
 * │  Browser State  │◄─────►│  Redis Storage  │◄──────►│  Redis Server       │
 * │  (API Layer)    │       │  Provider       │        │                     │
 * │                 │       │                 │        │                     │
 * └─────────────────┘       └─────────────────┘        └─────────────────────┘
 *                                 │    ▲
 *                                 │    │
 *                                 ▼    │
 *                           ┌─────────────────┐
 *                           │                 │
 *                           │  Temp Directory │
 *                           │  (ZIP Archive)  │
 *                           │                 │
 *                           └─────────────────┘
 *
 * Data Flow:
 * ---------
 *
 * Upload:
 * 1. Browser state calls upload() with a directory path containing profile files
 * 2. Directory is packaged into a single ZIP archive in a temporary location
 * 3. ZIP is encoded as base64 and stored in Redis at key: {prefix}:{userId}:{sessionId}
 * 4. Metadata is stored separately at key: {prefix}:{userId}:{sessionId}:metadata
 *
 * Download:
 * 1. Browser state calls download() with userId and sessionId
 * 2. Base64 data is fetched from Redis and decoded
 * 3. A temporary ZIP file is created from the decoded data
 * 4. ZIP is extracted to a directory that is returned to the browser state
 *
 * Session Management:
 * ------------------
 * - Sessions can have optional TTL (time-to-live) for automatic expiration
 * - listSessions() retrieves all sessions for a user by pattern matching
 * - deleteSession() removes both the session data and metadata
 * - Metadata includes timestamp and version information
 */

/**
 * Configuration options for Redis storage
 */
export interface RedisStorageOptions {
  /**
   * Redis server hostname
   * @default "localhost"
   */
  host: string;

  /**
   * Redis server port
   * @default 6379
   */
  port: number;

  /**
   * Optional Redis server password
   */
  password?: string;

  /**
   * Redis database number
   * @default 0
   */
  db?: number;

  /**
   * TLS/SSL configuration for secure Redis connections
   */
  tls?: RedisOptions["tls"];

  /**
   * Prefix for Redis keys to avoid collisions with other applications
   * @default "browserstate"
   */
  keyPrefix?: string;

  /**
   * Temporary directory for extracting and creating ZIP archives
   * @default os.tmpdir()
   */
  tempDir?: string;

  /**
   * Time-to-live in seconds for stored sessions
   * When specified, sessions will be automatically deleted after this time
   * @example 604800 // 7 days
   */
  ttl?: number;
}

/**
 * Validates that a ZIP entry path doesn't contain directory traversal attempts
 * to prevent ZIP slip vulnerability
 * @param entryPath - The path from the ZIP entry
 * @param targetDir - The directory where files will be extracted
 * @returns Whether the path is safe
 */
function isZipEntrySafe(entryPath: string, targetDir: string): boolean {
  // Normalize paths to handle different path formats
  const normalizedTarget = path.normalize(targetDir);
  const resolvedPath = path.resolve(normalizedTarget, entryPath);
  
  // Check if the resolved path is within the target directory
  // Use relative path to check if it goes outside the target
  return !path.relative(normalizedTarget, resolvedPath).startsWith('..');
}

/**
 * Redis storage provider for BrowserState
 *
 * This implementation stores browser profiles directly in Redis as ZIP archives.
 * Unlike cloud storage providers that store individual files, this approach:
 *
 * 1. Creates a complete ZIP archive of the entire browser profile directory
 * 2. Stores the archive as base64-encoded data in Redis
 * 3. Preserves the exact directory structure with all files intact
 * 4. Provides efficient storage through ZIP compression
 * 5. Maintains session metadata for tracking and management
 *
 * This implementation is ideal for:
 * - Microservices and containerized environments
 * - High-frequency, short-lived sessions
 * - Low-latency requirements
 * - When cloud storage dependencies are not desired
 *
 * Note: The Redis provider requires the 'ioredis' package to be installed.
 * It will be dynamically imported at runtime.
 */
export class RedisStorageProvider implements StorageProvider {
  private redis: RedisClient | null = null;
  private redisModulesLoaded = false;
  private keyPrefix: string;
  private tempDir: string;
  private ttl?: number;
  private options: RedisStorageOptions;

  /**
   * Creates a new Redis storage provider instance
   *
   * @param options - Redis connection and storage configuration
   */
  constructor(options: RedisStorageOptions) {
    this.keyPrefix = options.keyPrefix || "browserstate";
    
    // Validate keyPrefix format
    if (this.keyPrefix.includes(':')) {
      throw new Error("keyPrefix must not contain colons (:). The implementation automatically builds Redis keys in the format: {prefix}:{userId}:{sessionId}");
    }
    
    this.tempDir = options.tempDir || os.tmpdir();
    this.ttl = options.ttl;
    this.options = options;

    // Initialize with dynamic import (but don't throw if it fails)
    this.initClient().catch((error) => {
      if (process.env.NODE_ENV !== "production") {
        console.warn(
          "[Redis] Initialization failed, will retry on first usage:",
          error,
        );
      }
    });

    console.log(`[Redis] Storage initialized with ZIP compression`);
  }

  /**
   * Dynamically imports Redis module
   */
  private async initClient(): Promise<void> {
    if (this.redisModulesLoaded) return;

    try {
      // Import Redis using our module loader
      const Redis = await modules.redis.getModule();

      // Configure Redis connection
      const redisOptions: RedisOptions = {
        host: this.options.host,
        port: this.options.port,
        password: this.options.password,
        db: this.options.db,
        tls: this.options.tls,
        retryStrategy: (times: number) => {
          const delay = Math.min(times * 50, 2000);
          return delay;
        },
        maxRetriesPerRequest: 3,
        enableReadyCheck: true,
      };

      this.redis = new Redis(redisOptions) as unknown as RedisClient;
      this.redisModulesLoaded = true;
    } catch (error) {
      this.redis = null;

      // Always throw the error in development
      if (process.env.NODE_ENV !== "production") {
        throw error;
      }

      // In production, we'll throw when methods are called
    }
  }

  private async ensureInitialized(): Promise<void> {
    if (!this.redis) {
      await this.initClient();
      if (!this.redis) {
        throw new Error(
          "Failed to initialize Redis client. Please ensure ioredis is installed.",
        );
      }
    }
  }

  private getSessionKey(userId: string, sessionId: string): string {
    // Validate that userId and sessionId don't contain colons
    if (userId.includes(':')) {
      throw new Error("userId must not contain colons (:)");
    }
    if (sessionId.includes(':')) {
      throw new Error("sessionId must not contain colons (:)");
    }
    return `${this.keyPrefix}:${userId}:${sessionId}`;
  }

  private getMetadataKey(userId: string, sessionId: string): string {
    // Reuse validation from getSessionKey
    this.getSessionKey(userId, sessionId);
    return `${this.keyPrefix}:${userId}:${sessionId}:metadata`;
  }

  /**
   * Downloads a browser session from Redis
   *
   * This method:
   * 1. Creates a temporary directory for the session
   * 2. Fetches the ZIP archive from Redis
   * 3. Extracts the archive to recreate the browser profile
   *
   * If the session doesn't exist, an empty directory is returned
   * which can be used to create a new session.
   *
   * @param userId - User identifier
   * @param sessionId - Session identifier
   * @returns Path to the local directory containing session data
   */
  async download(userId: string, sessionId: string): Promise<string> {
    await this.ensureInitialized();
    if (!this.redis) {
      throw new Error("Redis client not initialized");
    }

    const sessionKey = this.getSessionKey(userId, sessionId);
    const metadataKey = this.getMetadataKey(userId, sessionId);

    // Create temporary directory for the session
    const userDataDir = path.join(
      this.tempDir,
      `browserstate-${userId}-${sessionId}`,
    );
    await fs.ensureDir(userDataDir);

    // Get session data (zipped folder) and metadata
    const [zipData, metadata] = await Promise.all([
      this.redis.get(sessionKey),
      this.redis.get(metadataKey),
    ]);

    if (!zipData) {
      console.log(
        `[Redis] No session data found for ${sessionId}, creating new directory`,
      );
      return userDataDir;
    }

    let sessionMetadata: SessionMetadata = {};
    try {
      if (metadata) {
        sessionMetadata = JSON.parse(metadata);
      }
    } catch (error) {
      console.warn(
        `[Redis] Error parsing metadata: ${error instanceof Error ? error.message : String(error)}`,
      );
    }

    console.log(
      `[Redis] Downloading session ${sessionId} from ${new Date(sessionMetadata.timestamp || 0).toISOString()}`,
    );

    // Create a temporary zip file
    const zipFilePath = path.join(
      this.tempDir,
      `${userId}-${sessionId}-${Date.now()}.zip`,
    );

    try {
      // Import extract-zip using our module loader
      const extractZip = await modules.extractZip.getModule();

      // Write the base64 data to a zip file
      await fs.writeFile(zipFilePath, Buffer.from(zipData, "base64"));

      // Extract the zip file to the user data directory
      console.log(`[Redis] Extracting zip to ${userDataDir}`);
      const options: ExtractOptions = { 
        dir: userDataDir,
        // Add onEntry callback to prevent ZIP slip vulnerability
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onEntry: (entry, _) => {
          // Ensure the entry's path doesn't escape the target directory (ZIP slip protection)
          const fileName = entry.fileName;
          if (!isZipEntrySafe(fileName, userDataDir)) {
            throw new Error(`Security risk: ZIP entry "${fileName}" is outside extraction directory`);
          }
        }
      };
      await extractZip(zipFilePath, options);

      // Clean up the temporary zip file
      await fs.remove(zipFilePath);

      console.log(`[Redis] Successfully extracted session data`);
      return userDataDir;
    } catch (error) {
      // Clean up temp zip file if it exists
      try {
        if (await fs.pathExists(zipFilePath)) {
          await fs.remove(zipFilePath);
        }
      } catch (cleanupError) {
        console.error(
          `[Redis] Error cleaning up temporary zip file: ${cleanupError instanceof Error ? cleanupError.message : String(cleanupError)}`,
        );
      }

      // Clean up user data directory to avoid partial extraction
      try {
        await fs.emptyDir(userDataDir);
      } catch (cleanupError) {
        console.error(
          `[Redis] Error cleaning up user data directory: ${cleanupError instanceof Error ? cleanupError.message : String(cleanupError)}`,
        );
      }

      console.error(
        `[Redis] Error extracting session data: ${error instanceof Error ? error.message : String(error)}`,
      );

      // Propagate the error instead of hiding it
      throw new Error(
        `Failed to extract session data for ${sessionId}: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }

  /**
   * Uploads a browser session to Redis
   *
   * This method:
   * 1. Creates a ZIP archive of the entire browser profile directory
   * 2. Encodes the archive as base64 and stores it in Redis
   * 3. Stores metadata about the session alongside the data
   *
   * If TTL is configured, both the session data and metadata
   * will expire after the specified time.
   *
   * @param userId - User identifier
   * @param sessionId - Session identifier
   * @param filePath - Path to the local directory containing session data
   */
  async upload(
    userId: string,
    sessionId: string,
    filePath: string,
  ): Promise<void> {
    await this.ensureInitialized();
    if (!this.redis) {
      throw new Error("Redis client not initialized");
    }

    if (!filePath) {
      throw new Error("Directory path is required");
    }

    const sessionKey = this.getSessionKey(userId, sessionId);
    const metadataKey = this.getMetadataKey(userId, sessionId);

    // Create a temporary file for the zip
    const zipFilePath = path.join(
      this.tempDir,
      `${userId}-${sessionId}-${Date.now()}.zip`,
    );

    console.log(
      `[Redis] Creating zip archive of session directory: ${filePath}`,
    );

    try {
      // Create a zip of the directory
      await this.zipDirectory(filePath, zipFilePath);

      // Get the zip file size for logging
      const stats = await fs.stat(zipFilePath);

      // Read the zip file as base64
      const zipData = await fs.readFile(zipFilePath, { encoding: "base64" });

      // Create metadata
      const metadata: SessionUploadMetadata = {
        timestamp: Date.now(),
        fileCount: 0, // We don't count individual files anymore
        version: "2.0", // Update version to indicate zip format
        encrypted: false, // Default to not encrypted
      };

      console.log(
        `[Redis] Uploading session ${sessionId} (${stats.size} bytes)`,
      );

      // Store in Redis with TTL if specified
      if (this.ttl) {
        await Promise.all([
          this.redis.setex(sessionKey, this.ttl, zipData),
          this.redis.setex(metadataKey, this.ttl, JSON.stringify(metadata)),
        ]);
      } else {
        await Promise.all([
          this.redis.set(sessionKey, zipData),
          this.redis.set(metadataKey, JSON.stringify(metadata)),
        ]);
      }

      console.log(`[Redis] Successfully uploaded session ${sessionId}`);

      // Clean up the temporary zip file
      await fs.remove(zipFilePath);
    } catch (error) {
      console.error(
        `[Redis] Error uploading session data: ${error instanceof Error ? error.message : String(error)}`,
      );
      throw error;
    }
  }

  /**
   * Creates a ZIP archive of a directory
   *
   * This helper method compresses an entire directory into a ZIP archive
   * with maximum compression level to minimize storage requirements.
   *
   * @param sourceDir - Directory to compress
   * @param outPath - Output path for the ZIP file
   * @returns Promise that resolves when the archive is created
   * @private
   */
  private async zipDirectory(
    sourceDir: string,
    outPath: string,
  ): Promise<void> {
    // Import archiver using our module loader
    const archiverModule = await modules.archiver.getModule();

    return new Promise((resolve, reject) => {
      const output = fs.createWriteStream(outPath);
      const archive = archiverModule("zip", {
        zlib: { level: 9 }, // Maximum compression
      });

      output.on("close", () => {
        console.log(`[Redis] ZIP archive created: ${archive.pointer()} bytes`);
        resolve();
      });

      archive.on("error", (err?: Error) => {
        reject(err);
      });

      archive.pipe(output);

      // Add the directory contents to the zip
      archive.directory(sourceDir, false);

      // Finalize the archive
      archive.finalize();
    });
  }

  /**
   * Lists all available sessions for a user
   *
   * This method:
   * 1. Searches for all Redis keys matching the user's pattern
   * 2. Filters out metadata keys to get only session identifiers
   * 3. Returns deduplicated session IDs
   *
   * @param userId - User identifier
   * @returns Array of session identifiers
   */
  async listSessions(userId: string): Promise<string[]> {
    await this.ensureInitialized();
    if (!this.redis) {
      throw new Error("Redis client not initialized");
    }

    // Validate userId
    if (userId.includes(':')) {
      throw new Error("userId must not contain colons (:)");
    }

    const pattern = `${this.keyPrefix}:${userId}:*`;
    const keys = await this.redis.keys(pattern);

    return keys
      .map((key: string) => {
        const match = key.match(new RegExp(`${this.keyPrefix}:${userId}:(.+?)(?::metadata)?$`));
        return match ? match[1] : "";
      })
      .filter(Boolean)
      .filter((key: string) => !key.includes(":"))
      .filter(
        (key: string, index: number, self: string[]) =>
          self.indexOf(key) === index,
      );
  }

  /**
   * Deletes a session and its metadata from Redis
   *
   * @param userId - User identifier
   * @param sessionId - Session identifier to delete
   */
  async deleteSession(userId: string, sessionId: string): Promise<void> {
    await this.ensureInitialized();
    if (!this.redis) {
      throw new Error("Redis client not initialized");
    }

    const sessionKey = this.getSessionKey(userId, sessionId);
    const metadataKey = this.getMetadataKey(userId, sessionId);

    await Promise.all([
      this.redis.del(sessionKey),
      this.redis.del(metadataKey),
    ]);
  }
}
